modules/product/src/main/java/com/opengamma/strata/product/cms/CmsLeg.java

```diff
@@ -0,0 +1,1058 @@
+import com.opengamma.strata.basics.schedule.PeriodicSchedule;
+import com.opengamma.strata.basics.schedule.Schedule;
+import com.opengamma.strata.basics.schedule.SchedulePeriod;
+import com.opengamma.strata.basics.value.ValueSchedule;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.product.swap.SwapIndex;
+import com.opengamma.strata.product.swap.type.IborRateSwapLegConvention;
+
+/**
+ * A CMS leg of a constant maturity swap (CMS) product. 
+ * <p>
+ * The class defines a single CMS leg for CMS or CMS cap/floor. 
+ * The CMS leg of the swap periodically pays coupons based on swap rate, the observed value of {@linkplain SwapIndex swap index}, 
+ * whereas that of CMS cap/floor is a set of call/put options on successive swap rates, i.e., CMS caplets/floorlets.
+ * <p>
+ * The periodic payments generated by {@link #expand()} are CMS coupons, CMS caplets or CMS floorlets following the 
+ * rule below.
+ * <p>
+ * {@code capSchedule} is used to represent strike values of individual caplets, whereas {@code floorSchedule} is used 
+ * to represent strike values of individual floorlets. 
+ * Thus at least one of {@code capSchedule} and {@code floorSchedule} must be empty. 
+ * If both the fields are null, the periodic payments in this leg are CMS coupons. 
+ */
+@BeanDefinition
+public final class CmsLeg
+    implements ImmutableBean, Serializable {
+
+  /**
+   * Whether the leg is pay or receive.
+   * <p>
+   * A value of 'Pay' implies that the resulting amount is paid to the counterparty.
+   * A value of 'Receive' implies that the resulting amount is received from the counterparty.
+   * Note that negative swap rates can result in a payment in the opposite direction to that implied by this indicator.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final PayReceive payReceive;
+  /**
+   * The periodic payment schedule.
+   * <p>
+   * This is used to define the periodic payment periods.
+   * These are used directly or indirectly to determine other dates in the leg.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final PeriodicSchedule periodicSchedule;
+  /**
+   * The swap index.
+   * <p>
+   * The swap rate to be paid is the observed value of this index. 
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final SwapIndex index;
+  /**
+   * The currency of the leg associated with the notional.
+   * <p>
+   * This is the currency of the leg and the currency that swap rate calculation is made in.
+   * The amounts of the notional are expressed in terms of this currency.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final Currency currency;
+  /**
+   * The notional amount, must be nonnegative. 
+   * <p>
+   * The notional amount applicable during the period.
+   * The currency of the notional is specified by {@code currency}.
+   */
+  @PropertyDefinition(validate = "ArgChecker.notNegative")
+  private final double notional;
```

Comment:
FpML allows the notional to vary in each payment period. If that can be priced, then this should be a `ValueSchedule`, but need to also consider if it implies notional exchanges...